src/pudl/workspace/setup.py

```diff
@@ -1,15 +1,114 @@
 
 import yaml
+from dotenv import load_dotenv
 
 import pudl.logging_helpers
 
 logger = pudl.logging_helpers.get_logger(__name__)
 
 
+def get_defaults(
+    input_dir: str | None = None,
+    output_dir: str | None = None,
+    yaml_file: IO | None = None,
+    default_pudl_yaml: Path | None = Path.home() / ".pudl.yml",
+) -> dict[str, str]:
+    """Derive PUDL workspace paths from specified input/output directories.
+
+    Determines input/output directory locations from YAML, then overrides with
+    env vars, then overrides with keywords passed in.
+
+    Input/output workspace roots can be the same directories.
+
+    Note: will update PUDL_OUTPUT and PUDL_INPUT env vars if they are
+    overridden by kwargs, so that Dagster configurations in child processes see
+    the updated configs when they read from env vars.
+
+    Args:
+        input_dir: equivalent to PUDL_INPUT environment variable, but overrides
+            that value. Derived paths treat the parent directory as the input
+            workspace root.
+        output_dir: equivalent to PUDL_OUTPUT environment variable, but
+            overrides that value. Derived paths treat the parent directory as
+            the output workspace root.
+        yaml_file: a buffer including the YAML configuration. The `pudl_in` and
+            `pudl_out` keys within this file correspond to the input/output
+            workspace roots directly, instead of through parents.
+
+    Returns:
+        dictionary with a variety of different paths where inputs/outputs are
+        to be found.
+    """
+    load_dotenv()
+
+    # Workaround for not having PUDL_* env vars in ReadTheDocs builds.
+    #
+    # They don't let you set env var through config files, and I'd rather
+    # have this in source control than go through some sort of web UI
+    #
+    # I don't like this any more than you do.
+    if os.getenv("READTHEDOCS"):
+        os.environ["PUDL_OUTPUT"] = str(Path("~/pudl-work/output").expanduser())
+        os.environ["PUDL_INPUT"] = str(Path("~/pudl-work/data").expanduser())
+
+    # read from YAML source
+    if yaml_file is not None:
+        yaml_settings = yaml.safe_load(yaml_file)
+    elif default_pudl_yaml and default_pudl_yaml.exists():
+        with default_pudl_yaml.open() as f:
+            yaml_settings = yaml.safe_load(f)
+    else:
+        yaml_settings = {}
+
+    # read from env vars
+    env_var_mapping = {
+        "pudl_in": os.getenv("PUDL_INPUT"),
+        "pudl_out": os.getenv("PUDL_OUTPUT"),
+    }
+    env_settings = {
+        key: str(Path(value).parent)
+        for key, value in env_var_mapping.items()
+        if value is not None
+    }
+
+    # read from params
+    kwarg_mapping = {"pudl_in": input_dir, "pudl_out": output_dir}
+    kwarg_settings = {
+        key: str(Path(value).parent)
+        for key, value in kwarg_mapping.items()
+        if value is not None
+    }
+
+    # Start with an empty settings, then override in order of precedence.
```

Comment:
This is generally meant to mirror the behavior of the Pydantic Settings classes so we can smoothly migrate into that system soon right? (which I am very excited about!)

Reply:
Yes!